File: AoC2025Day06.py
import sys, math

if __name__ == "__main__":
    if len(sys.argv) > 1:
        with open(sys.argv[1]) as inputFile:
            lines = inputFile.readlines()
            data = lines[:-1]
            ops = [x for x in lines[-1] if x == '+' or x =='*']
            partVals = []
            for line in data:
                nums = []
                val = 0
                for c in line.strip():
                    if c == ' ':
                        if val != 0:
                            nums.append(val)
                        val = 0
                    else:
                        val = val*10 + int(c)
                if val != 0:
                    nums.append(val)
                partVals.append(nums)
            
            part1, part2 = 0, 0
            for z in list(zip(*partVals, ops)):
                part1 += sum(z[:-1]) if z[-1] == '+' else math.prod(z[:-1])
        
            partVals.clear()
            zipped = list(zip(*data))
            for i in reversed(range(len(zipped)-1)):
                v = ''.join(zipped[i]).strip()
                if len(v):
                    partVals.append(int(v))
                if lines[-1][i] != ' ':
                    part2 += sum(partVals) if lines[-1][i] == '+' else math.prod(partVals)
                    partVals.clear()

            print(f"Part 1: {part1}\nPart 2: {part2}")
            

File: AoC2025Day06.fs
open System.IO

let ToNumbers (s:string) =
    let n, t = s.ToCharArray() |> Array.fold(fun (n,v) c -> if c=' ' || c='\n' then 
                                                                if v > 0L then (Array.append n [|v|], 0L) 
                                                                else (n,0L)
                                                            else (n, v*10L+(int64)(c-'0'))) ([||], 0L)
    if t > 0 then Array.append n [|t|] else n

let Part1 nums ops = 
    Array.map2(fun n o -> if o = '+' then Array.sum n else Array.fold (*) 1L n) nums ops |> Array.sum

let Part2 raw ops =
    let rawNums = Array.map(fun (s:string) -> s.ToCharArray()) raw |> Array.transpose
    let nums = Array.map(fun a -> Array.fold(fun v c -> if c = ' ' then v else v*10L+(int64)(c-'0')) 0L a) rawNums
    let grouped,work = Array.fold(fun (g,w) i -> if i = 0L then (Array.append g [|w|], [||])
                                                 else (g, Array.append w [|i|])) ([||], [||]) nums
    let groups = Array.append grouped [|work|]
    Part1 groups ops

[<EntryPoint>]
let main argv = 
    let raw = File.ReadAllLines(argv[0])
    let ops = raw[raw.Length-1].ToCharArray()|> Array.filter(fun c -> c ='+' || c='*')
    let p1Nums = Array.map ToNumbers raw[0..raw.Length-2] |> Array.transpose
    Part1 p1Nums ops |> printfn "Part 1: %A"
    Part2 raw[0..raw.Length-2] ops |> printfn "Part 2: %A"
    0

